Classify type with empty base as non-blittable

am11 opened the pull request:
Attempt to fix #46738.

am11 committed:
Add tests for MarshalUtils.IsBillitableType

am11 commented on src/coreclr/tools/aot/ILCompiler.TypeSystem.ReadyToRun.Tests/MarshalUtilsTests.cs:
```diff
@@ -0,0 +1,71 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
cc @jkoritzinsky, @MichalStrehovsky, I couldn't find MarshalUtils tests so I have added this. It may need to cover more cases?

AaronRobinsonMSFT commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/MarshalUtils.cs:
```diff
@@ -23,7 +23,10 @@ public static bool IsBlittableType(TypeDesc type)
-            if (hasNonTrivialParent && !IsBlittableType(baseType))
+            // Type is blittable only if parent is also blittable and is not empty.
+            if (hasNonTrivialParent && !IsBlittableType(baseType) ||
+                (!type.IsValueType && baseType is MetadataType baseMdType && !baseType.IsWellKnownType(WellKnownType.Object) &&
```
Why `!type.IsValueType`?

am11 replied:
Ah IsValueType is too strict, we just need `hasNonTrivialParent` to skip those well-known bases types from this check.

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/MarshalUtils.cs:
```diff
@@ -23,7 +23,10 @@ public static bool IsBlittableType(TypeDesc type)
-            if (hasNonTrivialParent && !IsBlittableType(baseType))
+            // Type is blittable only if parent is also blittable and is not empty.
+            if (hasNonTrivialParent && !IsBlittableType(baseType) ||
+                (hasNonTrivialParent && baseType is MetadataType baseMdType &&
```
Nit: `baseType` can be declared as `DefType` and then the cast to `MetadataType` will be unnecessary because the byte counts are available on `DefType`.

am11 replied:
Changed to `DefType`.

MichalStrehovsky replied:
Oh, I meant to change:

```
TypeDesc baseType = type.BaseType;
```

(above, so can't comment on that directly) into:

```
DefType baseType = type.BaseType;
```

am11 replied:
Cool. Updated and simplified the condition. There might be a better way to get the size without padding, but I tried several options and arithmetic operation seemed to be necessary to get LayoutInt.Zero for empty class having sequential layout. Approach also works for explicit layout.

MichalStrehovsky reviewed (commented):
I'll need to step through all of this in a debugger because I'm not sure what the logic should be. Layout-related things always require extra care. I don't have time/setup for that right now unfortunately (it's also why I just filed the issue instead of just fixing this :)).

Cc @dotnet/crossgen-contrib if someone has time to review this earlier.

am11 committed:
Classify type with empty base as non-blittable

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.TypeSystem.ReadyToRun.Tests/MarshalUtilsTests.cs:
```diff
@@ -0,0 +1,71 @@
+        [InlineData(WellKnownType.RuntimeMethodHandle)]
+        [InlineData(WellKnownType.TypedReference)]
+        [InlineData(WellKnownType.ByReferenceOfT)]
+        public void IsBlittableType_BilittableWellKnownTypes_ReturnsTrue(WellKnownType type) =>
```
Nullable, TypedReference and ByReferenceOfT should not be in this list.

- Nullable is an uninstantiated generic, and it doesn't make sense to ask this question.
- TypedReference and ByReferenceOfT should not be considered blittable at all as they hold byreference pointers which are not safe to marshal to native code under any circumstance.

davidwrighton commented on src/coreclr/tools/aot/ILCompiler.TypeSystem.ReadyToRun.Tests/MarshalUtilsTests.cs:
```diff
@@ -0,0 +1,71 @@
+        [InlineData(WellKnownType.Object)]
+        public void IsBlittableType_NonBilittableWellKnownTypes_ReturnsFalse(WellKnownType type) =>
+            Assert.False(MarshalUtils.IsBlittableType(_context.GetWellKnownType(type)));
+
```
Please add some tests where there is a single byte sized, 4 byte sized, and 8 byte field in the base type, and it still allows the derived type to be considered blittable.

davidwrighton commented on src/coreclr/tools/Common/TypeSystem/Interop/IL/MarshalUtils.cs:
```diff
@@ -18,12 +18,13 @@ public static bool IsBlittableType(TypeDesc type)
 
-            if (hasNonTrivialParent && !IsBlittableType(baseType))
+            // Type is blittable only if parent is also blittable and is not empty.
+            if (hasNonTrivialParent && (!IsBlittableType(baseType) || baseType.InstanceByteCountUnaligned - baseType.InstanceByteAlignment == LayoutInt.Zero))
```
The subtraction and compare against Zero doesn't seem right to me. What are the values that you're seeing in these cases? We may need to add a boolean flag like the native VM uses.

am11 replied:
I just read the code and debugged it with varied types to figure out that the only way in crossgen2 to get the size without padding is when InstanceByteCountUnaligned and InstanceByteAlignment are equal.

davidwrighton reviewed (changes requested):
@am11, thank you for looking into this. Please look at my comments and respond. Once you've done that I'll have a larger test pass run on this.

davidwrighton commented:
@am11  I see the new IsZeroSized bool property, but the implementation is quite confusing and feels like it is possibly fragile. Could you add a test that IsZeroSized returns false/true in a few interesting cases? In particular, if you could test the behavior of that api with both structs and classes that are empty/have a single field?

am11 committed:
Add tests for IsZeroSizedReferenceType

am11 commented:
@davidwrighton, admittedly, I am not an expert on crossgen2; but to my understanding, size calculation without padding: `_instanceByteCountUnaligned == _instanceByteAlignment` (or `_instanceByteCountUnaligned - _instanceByteAlignment == LayoutInt.Zero`) is true for empty reference type, and false for non-empty ones. This is the case for `LayoutKind.Explicit` and `LayoutKind.Sequential`.

I have added a comment and renamed the property to `IsZeroSizedReferenceType` as this logic is only applicable for classes.

am11 committed:
Rename local variables in tests

davidwrighton commented:
@am11 Thank you for indulging me on adding these various tests and the refactorings. IMO the way you are checking for empty in the property isn't a behavior that I believe we are certain to keep for all of time; however, I agree the behavior is correct at this time, and the tests are in place to ensure we don't break this behavior in the future (and if we do, all we'll have to do is fix the one property implementation).

davidwrighton commented:
@am11 Thank you for your contribution. We really appreciate it.
